tools/pyboard.py: Port recent patchs from mpremote serial transport, to add timeouts for non-responsive serial ports

dpgeorge opened the pull request:
### Summary

This applies #16513 and #16616 (made for mpremote) to `tools/pyboard.py`.  They add timeouts to the initial connection phase to a target board, to help with non-responsive serial ports.  Without these patches `pyboard.py` can hang indefinitely waiting for a board to respond.

### Testing

To be tested by @hmaerki on Octoprobe.

### Trade-offs and Alternatives

The test runners still use `pyboard.py`.  Eventually it would be good to change them to use `mpremote` as the backend, but that's a lot more effort.

dpgeorge committed:
tools/pyboard.py: Avoid initial blocking read in read_until().

This applies the mpremote commit 0d46e45a1f72ee61a153c41aaaf6c63818ccffb0
to pyboard.py.

If the target does not return any data then `read_until()` will block
indefinitely.  Fix this by making the initial read part of the general read
look, which always checks `inWaiting() > 0` before reading from the serial
device.

Also added the UART timeout to the constructor.  This is not currently used
but may be used as an additional safeguard.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/pyboard.py: Introduce timeout_overall for read_until().

This applies the mpremote commit 03fe9c55ea8a3ef8bede5547476f37d7fca20119
to pyboard.py.

The `timeout_overall` is used in `enter_raw_repl()`.  It prevents waiting
forever for a serial device that does not respond to the Ctrl-C/Ctrl-D/etc
commands and is constantly outputting data.

Signed-off-by: Damien George <damien@micropython.org>

hmaerki commented:
I reviewed this PR and it looks fine.
The testruns look much better - this is the newest one: https://reports.octoprobe.org/github_selfhosted_testrun_1/octoprobe_summary_report.html

I would propose to merge this PR!

<hr>
However, there is still a situation where the test time out with a `OSError: [Errno 5] Input/output error`. See:

* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@472b-ESP32_S3_DEVKIT/testresults.txt
* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@1830-LOLIN_C3_MINI/testresults.txt
* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@0c30-ESP32_C3_DEVKIT/testresults.txt
* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@2731-RPI_PICO2-RISCV/testresults.txt

I am still investigation why this error happens.

hmaerki commented:
In the summary, you write:

> This applies https://github.com/micropython/micropython/pull/16513 and https://github.com/micropython/micropython/pull/16616 (made for mpremote) to tools/pyboard.py.

It looks like a typo: 16513 should be 16531!

hmaerki commented:
However, there is still a situation where the test time out with a `OSError: [Errno 5] Input/output error`. See:

* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@472b-ESP32_S3_DEVKIT/testresults.txt
* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@1830-LOLIN_C3_MINI/testresults.txt
* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@0c30-ESP32_C3_DEVKIT/testresults.txt
* https://reports.octoprobe.org/github_selfhosted_testrun_1/RUN-TESTS_STANDARD_NATIVE@2731-RPI_PICO2-RISCV/testresults.txt

The problem was most probably that python multiprocessing terminated the processed and unpowered the tentacle. However, the terminating process still could see the powerdown and this is the ioctl error we see in above logs.

hmaerki commented:
https://reports.octoprobe.org/github_selfhosted_testrun_13/RUN-TESTS_EXTMOD_HARDWARE_NATIVE@2731-RPI_PICO2-RISCV/testresults.txt

This command still timeouts after 90s as pyboard hangs somehow.
@dpgeorge : Do you have an ideal where pyboard hangs? Or how to get a stacktrace at the moment of timeout..

dpgeorge committed:
tools/pyboard.py: Add write_timeout and catch errors in enter_raw_repl.

If the USB serial device locks up, then writes to that device can hang
forever.  That can make the test runner `tests/run-tests.py` lock up, among
other problems.

This commit introduces a 5 second write-timeout, and catches any OSError's
raised during `enter_raw_repl()`.  Now, if a USB serial device locks up,
`enter_raw_repl()` will eventually raise an exception.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I was unable to reproduce that error, even with exactly the same hardware (I think?) and using the `firmware.uf2` downloaded from that octoprobe run.

Anyway, I could force a similar error by making the board lock up during a test run.  That lead to a lock-up during `serial.write()`.  So I've now added a `write_timeout=5` parameter to the serial initialisation, and caught such errors during `enter_raw_repl()`.  My testing shows that this helps to catch such cases of the board locking up, although it's still quite slow because it will try to run every test after the lock-up and each one takes about 10 seconds.  Then it takes another 30 seconds or so to even exit `run-tests.py` due to the serial port having locked up.

This is not really an issue we can fix, because it's the OS that is blocking, waiting for the USB serial port to respond, and it never does respond.

dpgeorge commented:
@hmaerki I tried to trigger a run of this PR, but failed: https://reports.octoprobe.org/github_testbed_micropython_15

Can you please test this PR out (now with the added `write_timeout=5` bit)?

dpgeorge commented:
> From now on, all tests fail! It is probably the RPI_PICO2-RISCV which locks up!

Yes, I have seen this happen with this PR, if the device locks up then all tests from then on fail.  That's an improvement though!  Previously `run-tests.py` would just halt forever.  At least now it makes progress.
